File: python_kata/src/trees/binary_tree/pre_order_traversal.py
from typing import List
from trees.binary_tree.node import BinaryNode


def walk(curr: BinaryNode[int] | None, path: List[int]) -> List[int]:
    if not curr:
        return path

    # Pre
    path.append(curr.val)

    # Recursive
    walk(curr.left, path)
    walk(curr.right, path)

    # Post
    return path

def pre_order_traversal(head: BinaryNode[int]) -> List[int]:
    return walk(head, [])

File: typescript-kata/src/trees/binary-tree/preOrderTraversal.ts
import { BinaryNode, BinaryTree } from "./node";

function walk(curr: BinaryNode<number> | undefined, path: number[]): number[] {
  if (curr == undefined) {
    return path;
  }
  // Pre
  path.push(curr.val);

  // Recursive
  walk(curr.left, path);
  walk(curr.right, path);

  // Post
  return path;
}

export default function PreOrderTraversal(tree: BinaryTree<number>): number[] {
  if (!tree.root) {
    return [];
  }
  return walk(tree.root, []);
}
